src/pandroid/app/src/main/java/com/panda3ds/pandroid/view/controller/nodes/Joystick.java

```diff
@@ -0,0 +1,115 @@
+package com.panda3ds.pandroid.view.controller.nodes;
+
+import android.content.Context;
+import android.graphics.Canvas;
+import android.graphics.Color;
+import android.graphics.Paint;
+import android.graphics.drawable.Drawable;
+import android.util.AttributeSet;
+import androidx.annotation.NonNull;
+import androidx.appcompat.widget.AppCompatTextView;
+import com.panda3ds.pandroid.math.Vector2;
+import com.panda3ds.pandroid.view.controller.ControllerNode;
+import com.panda3ds.pandroid.view.controller.TouchEvent;
+import com.panda3ds.pandroid.view.controller.listeners.JoystickListener;
+
+public class Joystick extends BasicControllerNode implements ControllerNode {
+	private float stick_x = 0;
+	private float stick_y = 0;
+
+	private int size_width = 0;
+	private int size_height = 0;
+
+	private JoystickListener joystickListener;
+
+	public Joystick(Context context) { this(context, null); }
+
+	public Joystick(Context context, AttributeSet attrs) { this(context, attrs, 0); }
+
+	public Joystick(Context context, AttributeSet attrs, int defStyleAttr) {
+		super(context, attrs, defStyleAttr);
+
+		paint.setColor(Color.RED);
+		invalidate();
+	}
+
+	private final Paint paint = new Paint();
+
+	@Override
+	protected void onDraw(Canvas canvas) {
+		super.onDraw(canvas);
+	}
+
+	@Override
+	public void onDrawForeground(Canvas canvas) {
+		size_width = getWidth();
+		size_height = getHeight();
+
+		int analogIconSize = size_width - getPaddingLeft();
+
+		float middleIconSize = analogIconSize / 2.0F;
+		float middle = size_width / 2.0F;
+
+		float maxDistance = (middle - middleIconSize) * 0.9F;
+
+		float tx = maxDistance * stick_x;
+		float ty = maxDistance * stick_y;
+
+		float radius = Vector2.distance(0.0F, 0.0F, Math.abs(tx), Math.abs(ty));
+		radius = Math.min(maxDistance, radius);
+
+		double deg = Math.atan2(ty, tx) * (180.0 / Math.PI);
```

Comment:
It might be good to try and make this code avoid recalculating all of this if the circlepad state didn't change. A lot of this is non trivial to compute, especially the atan2 since arc tangent doesn't break down to a trivial Taylor polynomial.

Reply:
That is not a problem the onDraw only called if necessary ( or on call invalidade() )